### ISSUE TITLE
Set per connection schema

### ISSUE THREAD

**Issue by alex88** [go-gorm/gorm#2092]
### What version of Go are you using (`go version`)?
go version go1.11 darwin/amd64

### Which database and its version are you using?
postgres 10.5

### Please provide a complete runnable program to reproduce your issue. 
```go
package main

import (
	"github.com/jinzhu/gorm"
	_ "github.com/jinzhu/gorm/dialects/mssql"
	_ "github.com/jinzhu/gorm/dialects/mysql"
	_ "github.com/jinzhu/gorm/dialects/postgres"
	_ "github.com/jinzhu/gorm/dialects/sqlite"
)

var db *gorm.DB

func init() {
	var err error
	db, err = gorm.Open("postgres", "user=gorm password=gorm DB.name=gorm port=9920 sslmode=disable")
	if err != nil {
		panic(err)
	}
	db.LogMode(true)
}

func main() {
	db.Exec(fmt.Sprintf("SET search_path TO %s", "some schema"))
}
```

after running the above code the first DB connection uses the schema, however when using pooling all the other connections loses their default schema.
Is it possible to run such query for each pool connection? I know I can use other things but:
 - table name won't work since the schema is dynamic
 - use `DefaultTableNameHandler` but that also applies to external DB we use in our application logic

or is there another way?

**Comment by aaahrens** [comment]
Holy crap I'm actually dealing with this now

What I'm doing is using transactions then defering to commit or rollback. 

MY issue is that there are some automigrate problems going on with a many to many relation. I literally cannot automigrate a model with a foreign key if I set the search path, which is very odd.

**Comment by abdesselemaymen** [comment]
```
package database
// DB Database pointer
var DB *gorm.DB

var Schemas map[string]string

// Init schemas
func init() {
	Schemas = make(map[string]string)
}

// Connect establish
func Connect() {
	// utils.Println(">>>>>>>>>>>>>>>> CONNECTING >>>>>>>>>>>>>>>>>>")
	connectionString := fmt.Sprintf(
		"host=%s user=%s dbname=%s sslmode=%s password=%s",
		config.DBHost,
		config.DBUser,
		config.DBName,
		config.DBSSLMode,
		config.DBPassword,
	)

	DB, err = gorm.Open("postgres", connectionString)

	DB.DB().SetMaxIdleConns(50)
	DB.DB().SetMaxOpenConns(200)
	DB.DB().SetConnMaxLifetime(time.Minute * 1)

	DB.LogMode(false)

	// chang gorm timezone
	gorm.NowFunc = func() time.Time {
		return utils.GetCurrentTime()
	}
}

func SetSchema(schemaAddress uuid.UUID) *gorm.DB {
	tx := DB.Begin()
	tx.Exec("SET search_path TO " + Schemas[schemaAddress.String()] + " ,public")
	return tx
}

func CloseSchema(tx *gorm.DB) {
	tx.Exec("SET search_path TO public")
	tx.Commit()
}

func InitSchemas() {
	type Company struct {
		ID         uuid.UUID
		CompanyKey string
	}
	var companies []Company
	DB.Table("companies").Select("id,company_key").Scan(&companies)
	for _, company := range companies {
		Schemas[company.ID.String()] = company.CompanyKey
	}
}
```

**Comment by pedromorgan** [comment]
@abdesselemaymen Is above a workaround ??

**Comment by abdesselemaymen** [comment]
@pedromorgan I think the best solution is to add the schema name in the GORM query expl:
```go
datatabase.Model(User{}).Preload("Vendor").Schema("schema_name")...
```
result =>
```sql
set search_path to schema_name, public;
query...
```
"table name" can not solve the problem if the query contains multiple associations.

**Comment by mrclayman** [comment]
I think it's quite bad that gorm does not allow to set search path out-of-the-box in any way. I got bitten by this just now, too. We use QOR's admin module in our company and it looks like I am going to have to QOR's database handlers completely. (Up until now, I only set the search path beforehand and left the rest to QOR's original handlers, forgetting that database connections are pooled.) :unamused:
